blockchain/chain.go

```diff
@@ -1187,6 +1177,289 @@ func (b *BlockChain) FetchHeader(hash *chainhash.Hash) (wire.BlockHeader, error)
+		return nil, nil
+	}
+
+	// Grab a lock on the chain view to prevent it from changing due to a
+	// reorg while building the hashes.
+	b.bestChain.mtx.Lock()
+	defer b.bestChain.mtx.Unlock()
+
+	// When the requested start height is after the most recent best chain
+	// height, there is nothing to do.
+	latestHeight := b.bestChain.tip().height
+	if startHeight > latestHeight {
+		return nil, nil
+	}
+
+	// Limit the ending height to the latest height of the chain.
+	if endHeight > latestHeight+1 {
+		endHeight = latestHeight + 1
+	}
+
+	// Fetch as many as are available within the specified range.
+	hashes := make([]chainhash.Hash, 0, endHeight-startHeight)
+	for i := startHeight; i < endHeight; i++ {
+		hashes = append(hashes, b.bestChain.nodeByHeight(i).hash)
+	}
+	return hashes, nil
+}
+
+// locateBlocks returns the hashes of the blocks after the first known block in
+// the locator until the provided stop hash is reached, or up to the provided
+// max number of block hashes.
+//
+// See the comment on the exported function for more details on special cases.
+//
+// This function MUST be called with the chain state lock held (for reads).
+func (b *BlockChain) locateBlocks(locator BlockLocator, hashStop *chainhash.Hash, maxHashes uint32) []chainhash.Hash {
+	// There are no block locators so a specific block is being requested
+	// as identified by the stop hash.
+	stopNode := b.index.LookupNode(hashStop)
+	if len(locator) == 0 {
+		// No blocks with the stop hash were found so there is nothing
+		// to do.
+		if stopNode == nil {
+			return nil
+		}
+		return []chainhash.Hash{*hashStop}
+	}
+
+	// Find the most recent locator block hash in the main chain.  In the
+	// case none of the hashes in the locator are in the main chain, fall
+	// back to the genesis block.
+	startNode := b.bestChain.Genesis()
+	for _, hash := range locator {
+		node := b.index.LookupNode(hash)
+		if node != nil && b.bestChain.Contains(node) {
+			startNode = node
+			break
+		}
+	}
+
+	// Start at the block after the most most recently known block.  When
+	// there is no next block it means the most recently known block is
+	// the tip of the best chain, so there is nothing more to do.
+	startNode = b.bestChain.Next(startNode)
+	if startNode == nil {
+		return nil
+	}
+
+	// Calculate how many hashes will be returned.
+	total := uint32((b.bestChain.Tip().height - startNode.height) + 1)
+	if stopNode != nil && b.bestChain.Contains(stopNode) &&
+		stopNode.height >= startNode.height {
+
+		total = uint32((stopNode.height - startNode.height) + 1)
+	}
+	if total > maxHashes {
+		total = maxHashes
+	}
+
+	// Populate and return the found hashes.
+	hashes := make([]chainhash.Hash, 0, total)
+	for node := startNode; node != nil; node = b.bestChain.Next(node) {
+		hashes = append(hashes, node.hash)
+
+		// Stop once either the number of hashes exceeds the maximum or
+		// the stop hash has been reached.
+		if uint32(len(hashes)) >= maxHashes || node == stopNode {
```

Comment:
You can just check `uint32(len(hashes)) >= total` since total was calculated using the height of `stopNode`.